### PR TITLE
Move loss generating token counting to the dataloader

### PR DESCRIPTION
**Pull request by dakinggg**
The previous PR to count loss generating tokens caused a throughput regression because the token counting is on the critical path in the Composer trainer. This PR moves the token counting into the collator so that it can be hidden in the dataloader. It does this by counting tokens per row in the batch, using the existing token counting function (not the most efficient, but doesn't matter because this is not dataloader bottlenecked, and allows reusing the existing code).

This PR should not change token counts or loss, and will barely affect throughput for larger models, but have a significant impact for smaller models.

125m model pretraining throughput, loss, token count, before and after this PR:
Screenshot 2024-11-02 at 3 19 20 PM
Screenshot 2024-11-02 at 3 19 26 PM
Screenshot 2024-11-02 at 3 19 34 PM

llama finetune throughput, loss, token count, before and after this PR:

### CODE REVIEW

File: llmfoundry/data/contrastive_pairs/dataloader.py [hunk 1]
```diff
@@ -249,6 +249,11 @@ def collate_fn_without_labels(batch: list[Any]) -> dict[str, torch.Tensor]:
         processed_batch: dict[str, torch.Tensor] = collate_fn(batch)
         if 'labels' in processed_batch:
             del processed_batch['labels']
+
```
Comment (mvpatel2000) [inline llmfoundry/data/contrastive_pairs/dataloader.py:252]:
```suggestion
```
nit remove extra \n

File: llmfoundry/data/utils.py [hunk 2]
```diff
@@ -20,6 +20,55 @@
...
+                'input_ids': batch['input_ids'][row].unsqueeze(0),
+            }
+            if 'attention_mask' in batch:
+                row_batch['attention_mask'] = batch['attention_mask'][
+                    row].unsqueeze(0)
+            if 'labels' in batch:
+                row_batch['labels'] = batch['labels'][row].unsqueeze(0)
+            if 'decoder_attention_mask' in batch:
+                row_batch['decoder_attention_mask'] = batch[
+                    'decoder_attention_mask'][row].unsqueeze(0)
```
Comment (mvpatel2000) [inline llmfoundry/data/utils.py:54]: can we instead modify token_counting_func? It feels awkward to have a side-effect on other attributes here when it should just be appending new data

File: llmfoundry/data/utils.py [hunk 3]
```diff
@@ -20,6 +20,55 @@
 log = logging.getLogger(__name__)
 
 
+class LossGeneratingTokensCollatorWrapper:
+    """Collator wrapper to add loss generating token counts to batch."""
+
+    def __init__(
+        self,
+        base_collator: Callable,
+        token_counting_func: Callable[[Batch], Union[int, dict[str, int]]],
```
Comment (mvpatel2000) [inline llmfoundry/data/utils.py:29]: Do we need to pass the counting fn as an arg? given there's only one option and no way to pipe it through? Why not just call fn directly

**Review by mvpatel2000 — APPROVED** [review]
